src/components/Context.js

```diff
@@ -161,6 +178,76 @@ class ProductProvider extends Component {
             }
         })
     }
+
+    addPromotionToCart = couponCode => {
+        client.vouchers.get(couponCode, (error, result) => {
+            if (error) {
+                toast.error('Promotion not found')
+                return error
+            }
+            const voucher = result
+            console.log(voucher)
+            this.setState(() => {
+                return {
+                    promotionItems: voucher,
+                    promotionItemsNumber: 1,
+                    appliedCouponCode: couponCode
+                }
+            }, 
+            () => {
+                toast.success('Promotion applied')
+                this.countDiscount();
+            })
+        })
+    }
+    
+    removePromotionFromCart = () => {
+        this.setState(() => {
+            return {
+                promotionItems: {},
+                promotionItemsNumber: 0,
+                appliedCouponCode: '',
+                cartTotalAfterPromotion: 0
+            }
+        }, 
+        () => {
+            toast.success('Promotion removed')
+        })
+    }
+
+    countDiscount = () => {
```

Comment:
well, ideally, the discount is automatically calculated every time you modify e..g promotionItems or cartItems in your state.

A `reducer` could help.